README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# bssm

<!-- badges: start -->

[![Project Status: Active - The project has reached a stable, usable
state and is being actively
developed](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![Status at rOpenSci Software Peer
Review](https://badges.ropensci.org/489_status.svg)](https://github.com/ropensci/software-review/issues/489)
[![R-CMD-check](https://github.com/helske/bssm/workflows/R-CMD-check/badge.svg)](https://github.com/helske/bssm/actions)
[![Codecov test
coverage](https://codecov.io/gh/helske/bssm/branch/master/graph/badge.svg)](https://app.codecov.io/gh/helske/bssm?branch=master)
[![CRAN
version](http://www.r-pkg.org/badges/version/bssm)](https://CRAN.R-project.org/package=bssm)
[![downloads](https://cranlogs.r-pkg.org/badges/bssm)](https://cranlogs.r-pkg.org/badges/bssm)

<!-- badges: end -->

The `bssm` R package provides efficient methods for Bayesian inference
of state space models via particle Markov chain Monte Carlo and
importance sampling type weighted MCMC. Currently Gaussian, Poisson,
binomial, negative binomial, and Gamma observation densities with
linear-Gaussian state dynamics, as well as general non-linear Gaussian
models and discretely observed latent diffusion processes are supported.

For details, see

-   [The bssm paper on The R
    Journal](https://journal.r-project.org/archive/2021/RJ-2021-103/index.html),
-   [Package vignettes at CRAN](https://CRAN.R-project.org/package=bssm)
-   Paper on [Importance sampling type estimators based on approximate
    marginal Markov chain Monte
    Carlo](https://onlinelibrary.wiley.com/doi/abs/10.1111/sjos.12492)

There are also couple posters and a talk related to IS-correction
methodology and bssm package:

-   [UseR!2021 talk
    slides](https://jounihelske.netlify.app/talk/user2021/)  
-   [SMC 2017 workshop: Accelerating MCMC with an
    approximation](http://users.jyu.fi/~jovetale/posters/SMC2017)
-   [UseR!2017: Bayesian non-Gaussian state space models in
    R](http://users.jyu.fi/~jovetale/posters/user2017.pdf)

The `bssm` package was originally developed with the support of Academy
of Finland grants 284513, 312605, 311877, and 331817. Current
development is focused on increased usability. For recent changes, see
NEWS file.

### Citing the package

If you use the `bssm` package in publications, please cite the
corresponding R Journal paper:

Jouni Helske and Matti Vihola (2021). “bssm: Bayesian Inference of
Non-linear and Non-Gaussian State Space Models in R.” The R Journal
(2021) 13:2, pages 578-589.
<https://journal.r-project.org/archive/2021/RJ-2021-103/index.html>

## Installation

You can install the released version of bssm from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("bssm")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("helske/bssm")
```

Or from R-universe with

``` r
install.packages("bssm", repos = "https://helske.r-universe.dev")
```

## Example

Consider the daily air quality measurements in New Your from May to
September 1973, available in the `datasets` package. Let’s try to
predict the missing ozone levels by simple linear-Gaussian local linear
trend model with temperature and wind as explanatory variables (missing
response variables are handled naturally in the state space modelling
framework, however no missing values in covariates are normally
allowed);

``` r
library("bssm")
#> Warning: package 'bssm' was built under R version 4.1.3
#> 
#> Attaching package: 'bssm'
#> The following object is masked from 'package:base':
#> 
#>     gamma
library("dplyr")
#> Warning: package 'dplyr' was built under R version 4.1.3
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library("ggplot2")
#> Warning: package 'ggplot2' was built under R version 4.1.3
set.seed(1)

data("airquality", package = "datasets")

# Covariates as matrix. For complex cases, check out as_bssm function
xreg <- airquality %>% select(Wind, Temp) %>% as.matrix()

model <- bsm_lg(airquality$Ozone,
  xreg = xreg,  
  # Define priors for hyperparameters (i.e. not the states), see ?bssm_prior
  # Initial value followed by parameters of the prior distribution
  beta = normal_prior(rep(0, ncol(xreg)), 0, 1),
  sd_y = gamma_prior(1, 2, 0.01),
  sd_level = gamma_prior(1, 2, 0.01), 
  sd_slope = gamma_prior(1, 2, 0.01))

fit <- run_mcmc(model, iter = 20000, burnin = 5000)
fit
#> 
#> Call:
#> run_mcmc.lineargaussian(model = model, iter = 20000, burnin = 5000)
#> 
#> Iterations = 5001:20000
#> Thinning interval = 1
#> Length of the final jump chain = 3593
#> 
#> Acceptance rate after the burn-in period:  0.239
#> 
#> Summary for theta:
#> 
#>  variable       Mean          SE        SD        2.5%     97.5% ESS
#>  sd_level  6.3731836 0.113153715 2.8013937  1.52958636 12.403961 613
#>  sd_slope  0.3388712 0.010355574 0.2833955  0.04210885  1.070284 749
#>      sd_y 20.8618647 0.068145131 1.9369381 17.08728231 24.722309 808
#>      Temp  1.0265846 0.007497538 0.2064343  0.60226671  1.400436 758
#>      Wind -2.5183269 0.020978488 0.5764833 -3.68987992 -1.327578 755
#> 
#> Summary for alpha_154:
#> 
#>  variable time        Mean         SE        SD       2.5%     97.5%  ESS
#>     level  154 -28.3163054 0.69650977 20.132341 -69.271049 11.797133  835
#>     slope  154  -0.3740463 0.03683278  1.685733  -4.065499  2.830134 2094
#> 
#> Run time:
#>    user  system elapsed 
#>    1.00    0.03    1.03

obs <- data.frame(Time = 1:nrow(airquality),
  Ozone = airquality$Ozone) %>% filter(!is.na(Ozone))

pred <- fitted(fit, model)
pred %>%
  ggplot(aes(x = Time, y = Mean)) + 
  geom_ribbon(aes(ymin = `2.5%`, ymax = `97.5%`), 
    alpha = 0.5, fill = "steelblue") + 
  geom_line() + 
  geom_point(data = obs, 
    aes(x = Time, y = Ozone), colour = "Tomato") +
  theme_bw()
```

<img src="man/figures/README-example-1.png" width="100%" />

Same model but now assuming observations are from Gamma distribution:

``` r
model2 <- bsm_ng(airquality$Ozone,
  xreg = xreg,  
  beta = normal(rep(0, ncol(xreg)), 0, 1),
  distribution = "gamma",
  phi = gamma_prior(1, 2, 0.01),
  sd_level = gamma_prior(1, 2, 0.1), 
  sd_slope = gamma_prior(1, 2, 0.1))

fit2 <- run_mcmc(model2, iter = 20000, burnin = 5000, particles = 10)
fit2
#> 
#> Call:
#> run_mcmc.nongaussian(model = model2, iter = 20000, particles = 10, 
#>     burnin = 5000)
#> 
#> Iterations = 5001:20000
#> Thinning interval = 1
#> Length of the final jump chain = 3858
#> 
#> Acceptance rate after the burn-in period:  0.257
#> 
#> Summary for theta:
#> 
#>  variable         Mean           SE          SD          2.5%       97.5%  ESS
#>       phi  4.006977632 0.0159088062 0.536273508  3.0263444882  5.15527365 1136
#>  sd_level  0.057158663 0.0020138200 0.035366227  0.0083794202  0.14651419  308
#>  sd_slope  0.003894013 0.0001752319 0.003654978  0.0004250207  0.01374575  435
#>      Temp  0.052808820 0.0002404538 0.008701489  0.0353736458  0.06992423 1310
#>      Wind -0.057351094 0.0004338213 0.015411504 -0.0873384757 -0.02700112 1262
#>         SE_IS ESS_IS
#>  4.411840e-03  14611
#>  2.927386e-04  10591
#>  3.031489e-05   7766
#>  7.128104e-05  14485
#>  1.263047e-04  13905
#> 
#> Summary for alpha_154:
#> 
#>  variable time         Mean           SE         SD        2.5%      97.5%  ESS
#>     level  154 -0.200656509 0.0201721601 0.73134471 -1.62501396 1.24522802 1314
#>     slope  154 -0.002689176 0.0005121944 0.02289051 -0.04650504 0.04724173 1997
#>        SE_IS ESS_IS
#>  0.005987284   9458
#>  0.000191620   6448
#> 
#> Run time:
#>    user  system elapsed 
#>   11.20    0.11   11.25
```

Comparison:

``` r
pred2 <- fitted(fit2, model2)

bind_rows(list(Gaussian = pred, Gamma = pred2), .id = "Model") %>%
  ggplot(aes(x = Time, y = Mean)) + 
  geom_ribbon(aes(ymin = `2.5%`, ymax = `97.5%`, fill = Model), 
    alpha = 0.25) + 
  geom_line(aes(colour = Model)) + 
  geom_point(data = obs, 
    aes(x = Time, y = Ozone)) +
  theme_bw()
```

<img src="man/figures/README-compare-1.png" width="100%" />

Now let’s assume that we also want to use the solar radiation variable
as predictor for ozone. As it contains few missing values, we cannot use
it directly. As the number of missing time points is very small, simple
imputation would likely be acceptable, but let’s consider more another
approach. For simplicity, the slope terms of the previous models are now
omitted, and we focus on the Gaussian case. Let
![\\mu_t](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cmu_t "\mu_t")
be the true solar radiation at time
![t](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;t "t").
Now for ozone
![O_t](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;O_t "O_t")
we assume following model:

![O_t = D_t + \\alpha_t + \\beta_S \\mu_t + \\sigma\_\\epsilon \\epsilon_t](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;O_t%20%3D%20D_t%20%2B%20%5Calpha_t%20%2B%20%5Cbeta_S%20%5Cmu_t%20%2B%20%5Csigma_%5Cepsilon%20%5Cepsilon_t "O_t = D_t + \alpha_t + \beta_S \mu_t + \sigma_\epsilon \epsilon_t")  
![\\alpha\_{t+1} = \\alpha_t + \\sigma\_\\eta\\eta_t](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Calpha_%7Bt%2B1%7D%20%3D%20%5Calpha_t%20%2B%20%5Csigma_%5Ceta%5Ceta_t "\alpha_{t+1} = \alpha_t + \sigma_\eta\eta_t")  
![\\alpha_1 \\sim N(0, 100^2\\textrm{I})](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Calpha_1%20%5Csim%20N%280%2C%20100%5E2%5Ctextrm%7BI%7D%29 "\alpha_1 \sim N(0, 100^2\textrm{I})"),  
wheere
![D_t = \\beta X_t](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;D_t%20%3D%20%5Cbeta%20X_t "D_t = \beta X_t")
contains regression terms related to wind and temperature,
![\\alpha_t](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Calpha_t "\alpha_t")
is the time varying intercept term, and
![\\beta_S](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cbeta_S "\beta_S")
is the effect of solar radiation
![\\mu_t](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cmu_t "\mu_t").

Now for the observed solar radiation
![S_t](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;S_t "S_t")
we assume

![S_t = \\mu_t](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;S_t%20%3D%20%5Cmu_t "S_t = \mu_t")  
![\\mu\_{t+1} = \\mu_t + \\sigma\_\\xi\\xi_t,](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cmu_%7Bt%2B1%7D%20%3D%20%5Cmu_t%20%2B%20%5Csigma_%5Cxi%5Cxi_t%2C "\mu_{t+1} = \mu_t + \sigma_\xi\xi_t,")  

NEWS
bssm 2.0.1 (Release date: 2022-05-02)
==============
   * Fixed weights to one in case of non-linear model with mcmc_type="approx".
   * Adjusted tolerance of some testthat tests to comply with CRAN's MKL checks.

bssm 2.0.0 (Release date: 2021-11-26)
==============
   * Added a progress bar for run_mcmc.
   * Added a fitted method for extraction of summary statistics of posterior
     predictive distribution p(y_t | y_1, ..., y_n) for t = 1, ..., n.
   * Rewrote the summary method completely, which now returns data.frame. This
     also resulted in some changes in order of the function arguments.
   * The output of predict method is now a data frame with column weight
     corresponding to the IS-weights in case of IS-MCMC. Previously resampling
     was done internally, but now this is left for the user if needed
     (i.e. for drawing state trajectories).
   * The asymptotic_var and iact functions are now exported to users, and they
     also contain alternative methods based on the posterior package.
   * New function estimate_ess can be used to compute effective sample size
     from weighted MCMC.
   * Added compatibility with the posterior package by defining as_draws
     method for converting run_mcmc output to draws_df object.
   * New function check_diagnostics for quick glance of ESS and Rhat values.
   * Large number of new tests, and improved documentation with added examples.
   * Large number of internal tweaks so that the package complies with
     goodpractices package and Ropensci statistical software standards.

bssm 1.1.7-1 (Release date: 2021-09-21)
==============
   * Fixed an error in automatic tests due to lack of fixed RNG seed.

bssm 1.1.7 (Release date: 2021-09-20)
==============
   * Added a function cpp_example_model which can be used to extract and
     compile some non-linear and SDE models used in the examples and vignettes.
   * Added as_draws method for run_mcmc output so samples can be analysed using
     the posterior package.
   * Added more examples.
   * Fixed a tolerance of one MCMC test to pass the test on OSX as well.
   * Fixed a bug in iterated extended Kalman smoothing which resulted incorrect
     estimates.

bssm 1.1.6 (Release date: 2021-09-06)
==============
   * Cleaned some codes and added lots of tests in line with pkgcheck tests.
   * Fixed a bug in EKF-based particle filter which returned filtered estimates
     also in place of one-step ahead predictions.
   * Fixed a bug which caused an error in suggest_N for nlg_ssm.
   * Fixed a bug which caused incorrect sampling of smoothing distribution for
     ar1_lg model when predicting past or when using simulation smoother.
   * Fixed a bug which caused an error when predicting past values in
     multivariate time series case.
   * Fixed log-likelihood computation for gamma model with non-constant shape
     parameter when using (intermediate) Gaussian approximation.
   * Fixed sampling of negative binomial distribution in predict method, which
     used std::negative_binomial which converts non-integer phi to integer.
     Sampling now uses Gamma-Poisson mixture for simulation.


bssm 1.1.5 (Release date: 2021-06-14)
==============
   * Added explicit check for nsim > 0 in predict method as sample function
     works with missing argument causing crypting warnings later.
   * Updated drownings data until 2019 and changed the temperature variable
     to an average over three stations.
   * Improved checks for observations and distributions in model building.

bssm 1.1.4 (Release date: 2021-04-13)
==============
   * Better documentation for SV model, and changed ordering of arguments to
     emphasise the recommended parameterization.
   * Fixed predict method for SV model.
   * Removed parallelization in one example which failed on Solaris for some
     unknown reason.

bssm 1.1.3-2 (Release date: 2021-02-24)
==============
   * Fixed missing parenthesis causing compilation fail in case of no OpenMP
     support.
   * Added pandoc version >= 1.12.3 to system requirements.
   * Restructured C++ classes so no R structures are present in OpenMP regions.

bssm 1.1.3-1 (Release date: 2021-02-22)
==============
   * Fixed PM-MCMC and DA-MCMC for SDE models and added an example to `ssm_sde`.
   * Fixed the state covariance estimates of IS-MCMC, approx-MCMC, and
     Gaussian MCMC when output_type = "summary".
   * Fixed memory leaks due to uninitialized variables due to aborted particle
     filter.
   * Fixed numerical issues of multivariate normal density for nonlinear
     models.
   * Removed dependency on R::lchoose for safer parallel code.
   * Added vignette for SDE models.
   * Updated citation information and streamlined the main vignette.

bssm 1.1.2 (Release date: 2021-02-08)
==============
   * Changed the definition of D in ssm_ulg and ssm_ung, functions now accept
     D as scalar or vector as
     was originally intended.
   * Fixed a segfault issue with parallel state sampling in general
     ssm_ulg/mlg/ung/mng models caused by calls to R function inside parallel
     region.
   * Fixed a bug from version 1.0.0 in IS1 type sampling which actually lead
     to IS2 type sampling.
   * Fixed out-of-bounds error in IS3 sampling.
   * Fixed weight computations for multivariate nonlinear models in case of
     psi-APF in some border cases with non-standard H.
   * Removed Armadillo bound checks for efficiency gains.

bssm 1.1.1 (Release date: 2021-01-22)
==============

   * Added missing scaling for Gamma distribution in importance sampling
     weights for added numerical robustness.

R/bsm.R
#' Basic Structural (Time Series) Model
#'
#' Constructs a basic structural model with local level or local trend component
#' and seasonal component.
#'
#' @param y Vector or a \code{\link{ts}} object of observations.
#' @param sd_y Prior for the standard error of observation equation.
#' See \link[=uniform]{priors} for details.
#' @param sd_level  Prior for the standard error of the noise in level equation.
#' See\link[=uniform]{priors} for details. If missing, \code{sd_level} is fixed to zero.
#' @param sd_slope Prior for the standard error  of the noise in slope equation.
#' See\link[=uniform]{priors} for details. If missing, \code{sd_slope} is fixed to zero.
#' Ignored if \code{slope = FALSE}.
#' @param sd_seasonal Prior for the standard error of the noise in seasonal equation.
#' See\link[=uniform]{priors} for details. If missing, \code{sd_seasonal} is fixed to zero.
#' @param xreg Matrix containing covariates.
#' @param beta Prior for the regression coefficients.
#' @param period Length of the seasonal component i.e. the number of
#' observations per season. Default is \code{frequency(y)}.
#' @param slope Should the model contain the slope term. Default is \code{TRUE}.
#' If \code{FALSE}, local level component is used instead of local trend component.
#' @param seasonal Should the model contain seasonal term. Default is
#' \code{frequency(y) > 1}.
#' @param a1 Prior means for the initial states (level, slope, seasonals).
#' Defaults to vector of zeros.
#' @param P1 Prior covariance for the initial states (level, slope, seasonals).
#' Default is diagonal matrix with 1000 on the diagonal.
#' @return Object of class \code{bsm}.
#' @export
#' @examples
#'
#' prior <- uniform(0.1 * sd(log10(UKgas)), 0, 1)
#' model <- bsm(log10(UKgas), sd_y = prior,
#' sd_level =  prior, sd_slope =  prior, sd_seasonal =  prior)
#'
#' mcmc_out <- run_mcmc(model, n_iter = 5000)
#' summary(mcmc_out$theta)$stat
#' mcmc_out$theta[which.max(mcmc_out$posterior), ]
#' sqrt((fit <- StructTS(log10(UKgas), type = "BSM"))$coef)[c(4, 1:3)]
#'
bsm <- function(y, sd_y, sd_level, sd_slope, sd_seasonal,
  beta, xreg = NULL, period = frequency(y), a1, P1) {
 
  check_y(y)
  n <- length(y)
  
  if (is.null(xreg)) {
    xreg <- matrix(0, 0, 0)
    coefs <- numeric(0)
    beta <- NULL
  } else {
    
    if (missing(beta)) {
      stop("No prior defined for beta. ")
    }
    if(!is_prior(beta)) {
      stop("Prior for beta must be of class 'bssm_prior'.")
    }
    
    if (is.null(dim(xreg)) && length(xreg) == n) {
      xreg <- matrix(xreg, n, 1)
    }
    
    check_xreg(xreg, n)
    check_beta(beta$init, ncol(xreg))
    coefs <- beta$init
    if (is.null(colnames(xreg))) {
      colnames(xreg) <- paste0("coef_",1:ncol(xreg))
    }
    names(coefs) <- colnames(xreg)
  }
  
  notfixed <- c("y" = 1, "level" = 1, "slope" = 1, "seasonal" = 1)
  
  
  if (missing(sd_y)) {
    stop("Provide either prior or fixed value for sd_y.")
  } else {
    if (is_prior(sd_y)) {
      check_sd(sd_y$init, "y")
      H <- matrix(sd_y$init)
    } else {
      notfixed[1] <- 0
      check_sd(sd_y, "y")
      H <- matrix(sd_y)
    }
  }
 
  if (missing(sd_level)) {
    stop("Provide either prior or fixed value for sd_level.")
  } else {
    if (is_prior(sd_level)) {
      check_sd(sd_level$init, "level")
    } else {
      notfixed["level"] <- 0
      check_sd(sd_level, "level")
    }
  }
  
  if (missing(sd_slope)) {
    notfixed["slope"] <- 0
   slope <- FALSE
   sd_slope <- NULL
  } else {
    if (is_prior(sd_slope)) {
      check_sd(sd_slope$init, "sd_slope")
    } else {
      notfixed["slope"] <- 0
      check_sd(sd_slope, "sd_slope")
    }
    slope <- TRUE
  }

  if (missing(sd_seasonal)) {
      notfixed["seasonal"] <- 0
      seasonal_names <- NULL
      seasonal <- FALSE
      sd_seasonal <- NULL
  } else {
    if (period < 2) {
      stop("Period of seasonal component must be larger than 1. ")
    }
    if (is_prior(sd_seasonal)) {
      check_sd(sd_seasonal$init, "sd_seasonal")
    } else {
      notfixed["seasonal"] <- 0
      check_sd(sd_seasonal, "sd_seasonal")
    }
    seasonal <- TRUE
    seasonal_names <- paste0("seasonal_", 1:(period - 1))
  }

  npar_R <- 1L + as.integer(slope) + as.integer(seasonal)

  m <- as.integer(1L + as.integer(slope) + as.integer(seasonal) * (period - 1))

  if (missing(a1)) {
    a1 <- numeric(m)
  } else {
    if (length(a1) != m) {
      stop("Argument a1 must be a vector of length ", m)
    }
  }
  if (missing(P1)) {
    P1 <- diag(1e3, m)
  } else {
    if (is.null(dim(P1)) && length(P1) == 1L) {
      P1 <- matrix(P1)
    }
    if (!identical(dim(P1), c(m, m))) {
      stop("Argument P1 must be m x m matrix, where m = ", m)
    }
  }
 
  if (slope) {
    state_names <- c("level", "slope", seasonal_names)
  } else {
    state_names <- c("level", seasonal_names)
  }

  Z <- matrix(0, m, 1)
  Z[1, 1] <- 1
  if (seasonal) {
    Z[2 + slope, 1] <- 1
  }

  T <- matrix(0, m, m)
  T[1, 1] <- 1
  if (slope) {
    T[1:2, 2] <- 1
  }
  if (seasonal) {
    T[(2 + slope), (2 + slope):m] <- -1
    T[cbind(1 + slope + 2:(period - 1), 1 + slope + 1:(period - 2))] <- 1
  }

  R <- matrix(0, m, max(1, npar_R))

  if (notfixed["level"]) {
    R[1, 1] <- sd_level$init
  } else {
    R[1, 1] <- sd_level
  }
  if (slope) {
    if (notfixed["slope"]) {
      R[2, 2] <- sd_slope$init
    } else {
      R[2, 2] <- sd_slope
    }
  }
  if (seasonal) {
    if (notfixed["seasonal"]) {
      R[2 + slope, 2 + slope] <- sd_seasonal$init
    } else {
      R[2 + slope, 2 + slope] <- sd_seasonal
    }
  } 

  dim(H) <- 1
  dim(T) <- c(m, m, 1)
  dim(R) <- c(m, ncol(R), 1)

  names(a1) <- rownames(P1) <- colnames(P1) <- rownames(Z) <-
    rownames(T) <- colnames(T) <- rownames(R) <- state_names

 
  priors <- list(sd_y, sd_level, sd_slope, sd_seasonal, beta)
  names(priors) <- c("sd_y", "sd_level", "sd_slope", "sd_seasonal",names(coefs))
  priors <- priors[sapply(priors, is_prior)]

  structure(list(y = as.ts(y), Z = Z, H = H, T = T, R = R,
    a1 = a1, P1 = P1, xreg = xreg, coefs = coefs,
    slope = slope, seasonal = seasonal, period = period,
    fixed = as.integer(!notfixed), priors = priors), class = "bsm")
}

#' @method logLik bsm
#' @rdname logLik
#' @export
logLik.bsm <- function(object, ...) {
  bsm_loglik(object)
}

#' @method kfilter bsm
#' @rdname kfilter
#' @export
kfilter.bsm <- function(object, ...) {

  out <- bsm_filter(object)
  colnames(out$at) <- colnames(out$att) <- colnames(out$Pt) <-
    colnames(out$Ptt) <- rownames(out$Pt) <-
    rownames(out$Ptt) <- names(object$a1)
  out$at <- ts(out$at, start = start(object$y), frequency = object$period)
  out$att <- ts(out$att, start = start(object$y), frequency = object$period)
  out
}
#' @method fast_smoother bsm
#' @export
fast_smoother.bsm <- function(object, ...) {

  out <- bsm_fast_smoother(object)
  colnames(out) <- names(object$a1)
  ts(out, start = start(object$y), frequency = object$period)

}
#' @method sim_smoother bsm
#' @export
sim_smoother.bsm <- function(object, nsim = 1, seed = sample(.Machine$integer.max, size = 1), ...) {

  out <- bsm_sim_smoother(object, nsim, seed)

  rownames(out) <- names(object$a1)
  aperm(out, c(2, 1, 3))
}

#' @method smoother bsm
#' @export
smoother.bsm <- function(object, ...) {

  out <- bsm_smoother(object)

  colnames(out$alphahat) <- colnames(out$Vt) <- rownames(out$Vt) <- names(object$a1)
  out$alphahat <- ts(out$alphahat, start = start(object$y),
    frequency = object$period)
  out
}

#' @method run_mcmc bsm
#' @rdname run_mcmc_g
#' @param log_space Generate proposals for standard deviations in log-space. Default is \code{FALSE}.
#' @inheritParams run_mcmc.gssm
#' @export
run_mcmc.bsm <- function(object, n_iter, sim_states = TRUE, type = "full",
  n_burnin = floor(n_iter/2), n_thin = 1, gamma = 2/3,
  target_acceptance = 0.234, S, end_adaptive_phase = TRUE,
  log_space = FALSE,
  seed = sample(.Machine$integer.max, size = 1), ...) {

  type <- match.arg(type, c("full", "summary"))

  if(log_space) {
    stop("log_space = TRUE is under construction.")
  }

  if (missing(S)) {
    S <- diag(0.1 * pmax(0.1, abs(sapply(object$priors, "[[", "init"))), length(object$priors))
  }

  priors <- combine_priors(object$priors)

  out <- switch(type,
    full = {
      out <- bsm_run_mcmc(object,
        priors$prior_type, priors$params, n_iter,
        sim_states, n_burnin, n_thin, gamma, target_acceptance, S, seed, log_space, end_adaptive_phase)

      out$alpha <- aperm(out$alpha, c(2, 1, 3))
      colnames(out$alpha) <- names(object$a1)
      out
    },
    summary = {
      out <- bsm_run_mcmc_summary(object,
        priors$prior_type, priors$params, n_iter,
        n_burnin, n_thin, gamma, target_acceptance, S, seed,
        log_space, end_adaptive_phase)

      colnames(out$alphahat) <- colnames(out$Vt) <- rownames(out$Vt) <- names(object$a1)
      out$alphahat <- ts(out$alphahat, start = start(object$y),
        frequency = object$period)
      out
    })
  # if (log_space && (ncol(out$theta) - length(object$coefs)) > 0) {
  #   out$theta[, 1:n_sd_par] <- exp(out$theta[, 1:n_sd_par])
  # }
  names_ind <- !object$fixed & c(TRUE, TRUE, object$slope, object$seasonal)
  colnames(out$theta) <- rownames(out$S) <- colnames(out$S) <-
    c(c("sd_y", "sd_level", "sd_slope", "sd_seasonal")[names_ind],
      colnames(object$xreg))
  out$theta <- mcmc(out$theta, start = n_burnin + 1, thin = n_thin)
  out$call <- match.call()
  class(out) <- "mcmc_output"
  out
}

#' @method predict bsm
#' @rdname predict
#' @export
#' @examples
#' require("graphics")
#' y <- log10(JohnsonJohnson)
#' prior <- uniform(0.01, 0, 1)
#' model <- bsm(y, sd_y = prior, sd_level = prior,
#'   sd_slope = prior, sd_seasonal = prior)
#'
#' pred1 <- predict(model, n_iter = 5000, n_ahead = 8)
#' pred2 <- predict(StructTS(y, type = "BSM"), n.ahead = 8)
#'
#' ts.plot(pred1$mean, pred1$intervals[,-2], pred2$pred +
#' cbind(0, -qnorm(0.95) * pred2$se, qnorm(0.95) * pred2$se),
#'   col = c(1, 1, 1, 2, 2, 2))
#'
predict.bsm <- function(object, n_iter, newdata = NULL,
  n_ahead = 1, interval = "response", probs = c(0.05, 0.95),
  method = "quantile", return_MCSE = TRUE,
  n_burnin = floor(n_iter/2),
  n_thin = 1, gamma = 2/3, target_acceptance = 0.234, S,
  seed = sample(.Machine$integer.max, size = 1), ...) {

  interval <- pmatch(interval, c("mean", "response"))
  method <- match.arg(method, c("parametric", "quantile"))

  if (missing(S)) {
    S <- diag(0.1 * pmax(0.1, abs(sapply(object$priors, "[[", "init"))), length(object$priors))
  }

  priors <- combine_priors(object$priors)


  endtime <- end(object$y) + c(0, n_ahead)
  y_orig <- object$y
  object$y <- c(object$y, rep(NA, n_ahead))

  if (length(object$coefs) > 0) {
    if (is.null(newdata) || nrow(newdata) != n_ahead ||
        ncol(newdata) != length(object$coefs)) {
      stop("Model contains regression part but newdata is missing or its dimensions do not match with n_ahead and length of beta. ")
    }
    object$xreg <- rbind(object$xreg, newdata)
  }
  probs <- sort(unique(c(probs, 0.5)))
  if (method == "parametric") {
    out <- bsm_predict(object,
      priors$prior_types, priors$params, n_iter,
      n_burnin, n_thin, gamma, target_acceptance, S, n_ahead, interval,
      probs, seed, FALSE)

    if (return_MCSE) {
      ses <- matrix(0, n_ahead, length(probs))

      nsim <- nrow(out$y_mean)

      for (i in 1:n_ahead) {
        for (j in 1:length(probs)) {
          pnorms <- pnorm(q = out$intervals[i, j], out$y_mean[, i], out$y_sd[, i])
          eff_n <-  effectiveSize(pnorms)
          ses[i, j] <- sqrt((sum((probs[j] - pnorms) ^ 2) / nsim) / eff_n) /
            sum(dnorm(x = out$intervals[i, j], out$y_mean[, i], out$y_sd[, i]) / nsim)
        }
      }

      pred <- list(y = y_orig, mean = ts(colMeans(out$y_mean), end = endtime, frequency = object$period),
        intervals = ts(out$intervals, end = endtime, frequency = object$period,
          names = paste0(100 * probs, "%")),
        MCSE = ts(ses, end = endtime, frequency = object$period,
          names = paste0(100 * probs, "%")))
    } else {
      pred <- list(y = y_orig, mean = ts(colMeans(out$y_mean), end = endtime, frequency = object$period),
        intervals = ts(out$intervals, end = endtime, frequency = object$period,
          names = paste0(100 * probs, "%")))
    }
  } else {
    out <- bsm_predict2(object, priors$prior_types, priors$params, n_iter,
      n_burnin, n_thin, gamma, target_acceptance, S, n_ahead, interval,
      seed, FALSE)

    pred <- list(y = y_orig, mean = ts(rowMeans(out), end = endtime, frequency = object$period),
      intervals = ts(t(apply(out, 1, quantile, probs, type = 8)), end = endtime, frequency = object$period,
        names = paste0(100 * probs, "%")))
  }
  class(pred) <- "predict_bssm"
  pred
}

#' @method particle_filter bsm
#' @rdname particle_filter
#' @export
particle_filter.bsm <- function(object, nsim,
  seed = sample(.Machine$integer.max, size = 1), ...) {

  out <- bsm_particle_filter(object, nsim, seed)

  rownames(out$alpha) <- names(object$a1)
  out$alpha <- aperm(out$alpha, c(2, 1, 3))
  out
}

#' @method particle_smoother bsm
#' @rdname particle_smoother
#' @export
particle_smoother.bsm <- function(object, nsim, method = "fs",
  seed = sample(.Machine$integer.max, size = 1), ...) {

  method <- match.arg(method, c("fs", "fbs"))
  out <- bsm_particle_smoother(object, nsim, seed, method == "fs")

  rownames(out$alpha) <- names(object$a1)
  out$alpha <- aperm(out$alpha, c(2, 1, 3))
  out
}

#' @method particle_simulate bsm
#' @rdname particle_simulate
#' @export
particle_simulate.bsm <- function(object, nsim, nsim_store = 1,
  seed = sample(.Machine$integer.max, size = 1), ...) {

  out <- bsm_backward_simulate(object, nsim, seed, nsim_store)

  rownames(out$alpha) <- names(object$a1)
  out$alpha <- aperm(out$alpha, c(2, 1, 3))
  out
}
